modules/samples/authenticators/components/org.wso2.carbon.identity.sample.extension.authenticators/src/main/java/org/wso2/carbon/identity/sample/extension/auth/SampleRetinaAuthenticator.java

```diff
@@ -0,0 +1,135 @@
+
+        return APP_URL;
+    }
+
+    @Override
+    public String getContextIdentifier(HttpServletRequest request) {
+
+        String identifier = request.getParameter("sessionDataKey");
+        return identifier;
+    }
+
+    @Override
+    public String getName() {
+
+        return "SampleRetinaAuthenticator";
+    }
+
+    @Override
+    public String getFriendlyName() {
+
+        return "Sample Retina Authenticator";
+    }
+
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest httpServletRequest,
+                                                 HttpServletResponse httpServletResponse,
+                                                 AuthenticationContext authenticationContext)
+            throws AuthenticationFailedException {
+
+    }
+
+    @Override
+    public String getClaimDialectURI() {
+
+        return null;
+    }
+
+    @Override
+    public List<Property> getConfigurationProperties() {
+
+        List<Property> configProperties = new ArrayList<>();
+
+        Property appUrl = new Property();
+        appUrl.setName(APP_URL);
+        appUrl.setValue(IdentityUtil
+                .getServerURL("sample-auth", true, true) + "/retina.jsp");
+        appUrl.setDisplayName("Retina Sample URL");
+        appUrl.setRequired(true);
+        appUrl.setDescription("Enter sample Retina url value.");
+        appUrl.setDisplayOrder(0);
+        configProperties.add(appUrl);
+        return configProperties;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response,
+                                                 AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String loginPage = IdentityUtil
+                .getServerURL("sample-auth", true, true) + "/retina.jsp";
+
+        String queryParams =
+                FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                        context.getCallerSessionKey(),
+                        context.getContextIdentifier());
+        try {
+            String retryParam = "";
+            if (context.isRetrying()) {
+                retryParam = "&authFailure=true&authFailureMsg=login.fail.message";
+            }
+            String callbackUrl = IdentityUtil
+                    .getServerURL(FrameworkConstants.COMMONAUTH, true, true);
+            callbackUrl = callbackUrl + "?sessionDataKey=" + context.getContextIdentifier();
+            String encodedUrl = URLEncoder.encode(callbackUrl, StandardCharsets.UTF_8.name());
+
+            response.sendRedirect(response.encodeRedirectURL(loginPage + ("?" + queryParams)) +
+                    "&callbackUrl=" + encodedUrl + "&authenticators=SampleRetinaAuthenticator:" + "LOCAL" + retryParam);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException(e.getMessage(), e);
```

Comment:
e.getMessage is anyway there within the exception. Hence you should add a more specific message here.